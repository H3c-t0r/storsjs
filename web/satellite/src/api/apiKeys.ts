// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import apollo from '@/utils/apolloManager';
import gql from 'graphql-tag';

export async function fetchAPIKeys(projectID: string) {
    let result: RequestResponse<any[]> = {
        errorMessage: '',
        isSuccess: false,
        data: []
    };

    try {
        let response: any = await apollo.query({
            query: gql(
                `query {
                    project(
                        id: "${projectID}",
                    ) {
                        apiKeys {
                            id,
                            name,
                            createdAt
                        }
                    }
                }`
            ),
            fetchPolicy: 'no-cache'
        });

        if (response.errors) {
            result.errorMessage = response.errors[0].message;
        } else {
            result.isSuccess = true;
            result.data = response.data.project.apiKeys;
        }
    } catch (e) {
        result.errorMessage = e.message;
    }

    return result;
}

export async function createAPIKey(projectID: string, name: string) {
    let result: RequestResponse<any> = {
        errorMessage: '',
        isSuccess: false,
        data: null
    };

    try {
        let response: any = await apollo.mutate({
            mutation: gql(
                `mutation {
                    createAPIKey(
                        projectID: "${projectID}",
                        name: "${name}"
                    ) {
                        key,
                        keyInfo {
                            id,
                            name,
                            createdAt
                        }
                    }
                }`
            ),
            fetchPolicy: 'no-cache'
        });

        if (response.errors) {
            result.errorMessage = response.errors[0].message;
        } else {
            console.log('response', response)
            result.isSuccess = true;
            result.data = {
                key: response.data.createAPIKey.key,
                keyInfo: response.data.createAPIKey.keyInfo
            }
        }
    } catch (e) {
        result.errorMessage = e.message;
    }

    return result;
}

export async function deleteAPIKey(id: string) {
    let result: RequestResponse<any> = {
        errorMessage: '',
        isSuccess: false,
        data: null
    };

    try {
        let response: any = await apollo.mutate({
            mutation: gql(
                `mutation {
                    deleteAPIKey(id: "${id}") {
                        id
                    }
                }`
            ),
            fetchPolicy: 'no-cache'
        });

        if (response.errors) {
            result.errorMessage = response.errors[0].message;
        } else {
            result.isSuccess = true;
            result.data = response.data.deleteAPIKey;
        }
    } catch (e) {
        result.errorMessage = e.message;
    }

    return result;
}