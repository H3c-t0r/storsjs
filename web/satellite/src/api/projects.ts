// Copyright (C) 2018 Storj Labs, Inc.
// See LICENSE for copying information.

import apollo from '@/utils/apolloManager';
import gql from 'graphql-tag';

// Performs graqhQL request.
// Throws an exception if error occurs
export async function createProject(project: Project): Promise<any> {
	let response = await apollo.mutate(
		{
			mutation: gql(`
                mutation {
                    createProject(
                        input: {
                            name: "${project.name}",
                            description: "${project.description}",
                            companyName: "${project.companyName}",
                            isTermsAccepted: ${project.isTermsAccepted},
                        }
                    ) {id}
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot create project');

		return null;
	}

	return response;
}

// Performs graqhQL request for fetching all projects of current user.
export async function fetchProjects(): Promise<any> {
	let response = await apollo.query(
		{
			query: gql(`
                query {
                    myProjects{
                        name
                        id
                        description
                        createdAt            
                        ownerName           
                        companyName           
                        isTermsAccepted
                    }
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot fetch projects');

		return null;
	}

	return response;
}

// Performs graqhQL request for updating selected project description
export async function updateProject(projectID: string, description: string): Promise<any> {
	let response = await apollo.mutate(
		{
			mutation: gql(`
                mutation {
                    updateProjectDescription(
                        id: "${projectID}",
                        description: "${description}"
                    )
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot update project');

		return null;
	}

	return response;
}

// Performs graqhQL request for deleting selected project
export async function deleteProject(projectID: string): Promise<any> {
	let response = await apollo.mutate(
		{
			mutation: gql(`
                mutation {
                    deleteProject(
                        id: "${projectID}"
                    )
                }`
			),
			fetchPolicy: 'no-cache',
		}
	);

	if (!response) {
		// TODO: replace with popup in future
		console.error('cannot delete project');

		return null;
	}

	return response;
}
