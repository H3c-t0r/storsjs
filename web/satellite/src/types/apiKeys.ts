// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Exposes all apiKey-related functionality
 */
export interface ApiKeysApi {
    /**
     * Fetch apiKey
     *
     * @returns ApiKey
     * @throws Error
     */
    get(projectId: string): Promise<ApiKey[]>;
    create(projectId: string, name: string): Promise<ApiKey>;
    delete(ids: string[]): Promise<null>;
}

/**
 * ApiKey class holds info for ApiKey entity.
 */
export class ApiKey {
    public id: string;
    public secret: string;
    public name: string;
    public createdAt: string;
    public isSelected: boolean = false;

    constructor(id: string, name: string, createdAt: string, secret: string) {
        this.id = id || '';
        this.name = name || '';
        this.createdAt = createdAt || '';
        this.secret = secret || '';

        this.isSelected = false;
    }

    public formattedName(): string {
        let name = this.name;

        if (name.length < 12) {
            return name;
        }

        return name.slice(0, 12) + '...';
    }

    public getDate(): string {
        if (!this.createdAt) {
            return '';
        }

        return new Date(this.createdAt).toLocaleDateString();
    }
}
