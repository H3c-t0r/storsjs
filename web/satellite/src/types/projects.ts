// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

/**
 * Exposes all project-related functionality
 */
export interface ProjectsApi {
    /**
     * Creates project
     *
     * @param createProjectModel - contains project information
     * @throws Error
     */
    create(createProjectModel: CreateProjectModel): Promise<Project>;
    /**
     * Fetch projects
     *
     * @returns Project[]
     * @throws Error
     */
    get(): Promise<Project[]>;
    /**
     * Update project
     *
     * @param projectId - project ID
     * @param description - project description
     * @returns Project[]
     * @throws Error
     */
    update(projectId: string, description: string): Promise<void>;
    /**
     * Delete project
     *
     * @param projectId - project ID
     * @throws Error
     */
    delete(projectId: string): Promise<void>;
}

// Project is a type, used for creating new project in backend
export class Project {
    public id: string;

    public name: string;
    public description: string;
    public createdAt: string;
    public ownerId: string;

    public isSelected: boolean;

    public constructor(id: string = '', name: string = '', description: string = '', createdAt: string = '', ownerId: string = '', isSelected: boolean = false) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.createdAt = createdAt;
        this.isSelected = isSelected;
        this.ownerId = ownerId;
    }
}

// UpdateProjectModel is a type, used for updating project description
export class UpdateProjectModel {
    public id: string;
    public description: string;

    public constructor(id: string, description: string) {
        this.id = id;
        this.description = description;
    }
}

// CreateProjectModel is a type, used for creating project
export class CreateProjectModel {
    public name: string;
    public description: string;
}
