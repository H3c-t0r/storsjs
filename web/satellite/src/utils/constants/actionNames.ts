// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

export const NOTIFICATION_ACTIONS = {
    SUCCESS: 'success',
    ERROR: 'error',
    NOTIFY: 'notify',
    WARNING: 'WARNING',
    DELETE: 'deleteNotification',
    PAUSE: 'pauseNotification',
    RESUME: 'resumeNotification',
    CLEAR: 'clearNotifications',
};
