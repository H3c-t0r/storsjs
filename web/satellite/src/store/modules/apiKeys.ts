// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { ApiKey, ApiKeyCursor, ApiKeyOrderBy, ApiKeysApi, ApiKeysPage } from '@/types/apiKeys';
import { SortDirection } from '@/types/common';

export const API_KEYS_MUTATIONS = {
    SET_PAGE: 'setApiKeys',
    TOGGLE_SELECTION: 'toggleApiKeysSelection',
    CLEAR_SELECTION: 'clearApiKeysSelection',
    CLEAR: 'clearApiKeys',
    CHANGE_SORT_ORDER: 'changeApiKeysSortOrder',
    CHANGE_SORT_ORDER_DIRECTION: 'changeApiKeysSortOrderDirection',
    SET_SEARCH_QUERY: 'setApiKeysSearchQuery',
    SET_PAGE_NUMBER: 'setApiKeysPage',
};

const {
    SET_PAGE,
    TOGGLE_SELECTION,
    CLEAR_SELECTION,
    CLEAR,
    CHANGE_SORT_ORDER,
    CHANGE_SORT_ORDER_DIRECTION,
    SET_SEARCH_QUERY,
    SET_PAGE_NUMBER,
} = API_KEYS_MUTATIONS;

class ApiKeysState {
    public cursor: ApiKeyCursor = new ApiKeyCursor();
    public page: ApiKeysPage = new ApiKeysPage();
}

/**
 * creates apiKeys module with all dependencies
 *
 * @param api - apiKeys api
 */
export function makeApiKeysModule(api: ApiKeysApi): StoreModule<ApiKeysState> {
    return {
        state: new ApiKeysState(),
        mutations: {
            [SET_PAGE](state: ApiKeysState, page: ApiKeysPage) {
                state.page = page;
            },
            [SET_PAGE_NUMBER](state: ApiKeysState, pageNumber: number) {
                state.cursor.page = pageNumber;
            },
            [SET_SEARCH_QUERY](state: ApiKeysState, search: string) {
                state.cursor.search = search;
            },
            [CHANGE_SORT_ORDER](state: ApiKeysState, order: ApiKeyOrderBy) {
                state.cursor.order = order;
            },
            [CHANGE_SORT_ORDER_DIRECTION](state: ApiKeysState, direction: SortDirection) {
                state.cursor.orderDirection = direction;
            },
            [TOGGLE_SELECTION](state: ApiKeysState, apiKeyID: string) {
                state.page.apiKeys = state.page.apiKeys.map((apiKey: ApiKey) => {
                    if (apiKey.id === apiKeyID) {
                        apiKey.isSelected = !apiKey.isSelected;
                    }

                    return apiKey;
                });
            },
            [CLEAR_SELECTION](state: ApiKeysState) {
                state.page.apiKeys = state.page.apiKeys.map((apiKey: ApiKey) => {
                    apiKey.isSelected = false;

                    return apiKey;
                });
            },
            [CLEAR](state: ApiKeysState) {
                state.cursor = new ApiKeyCursor();
                state.page = new ApiKeysPage();
            },
        },
        actions: {
            fetchApiKeys: async function ({commit, rootGetters, state}, pageNumber: number): Promise<ApiKeysPage> {
                const projectId = rootGetters.selectedProject.id;
                commit(SET_PAGE_NUMBER, pageNumber);

                const apiKeys = await api.get(projectId, state.cursor);
                commit(SET_PAGE, apiKeys);

                return apiKeys;
            },
            createApiKey: async function ({commit, rootGetters}: any, name: string): Promise<ApiKey> {
                const projectId = rootGetters.selectedProject.id;

                const apiKey = await api.create(projectId, name);

                return apiKey;
            },
            deleteApiKey: async function({commit}: any, ids: string[]): Promise<void> {
                return await api.delete(ids);
            },
            setApiKeysSearchQuery: function ({commit}, search: string) {
                commit(SET_SEARCH_QUERY, search);
            },
            setApiKeysSortingBy: function ({commit}, order: ApiKeyOrderBy) {
                commit(CHANGE_SORT_ORDER, order);
            },
            setApiKeysSortingDirection: function ({commit}, direction: SortDirection) {
                commit(CHANGE_SORT_ORDER_DIRECTION, direction);
            },
            toggleApiKeySelection: function ({commit}, apiKeyID: string): void {
                commit(TOGGLE_SELECTION, apiKeyID);
            },
            clearApiKeySelection: function ({commit}): void {
                commit(CLEAR_SELECTION);
            },
            clearApiKeys: function ({commit}): void {
                commit(CLEAR);
            },
        },
        getters: {
            selectedApiKeys: (state: ApiKeysState) => state.page.apiKeys.filter((key: ApiKey) => key.isSelected),
            apiKeys: function (state: ApiKeysState): ApiKey[] {
                return state.page.apiKeys;
            },
        },
    };
}
