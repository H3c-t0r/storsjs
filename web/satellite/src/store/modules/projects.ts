// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { StoreModule } from '@/store';
import { CreateProjectModel, Project, ProjectsApi, UpdateProjectModel } from '@/types/projects';

export const PROJECTS_ACTIONS = {
    FETCH: 'fetchProjects',
    CREATE: 'createProject',
    SELECT: 'selectProject',
    UPDATE: 'updateProject',
    DELETE: 'deleteProject',
    CLEAR: 'clearProjects',
};

export const PROJECTS_MUTATIONS = {
    ADD: 'CREATE_PROJECT',
    REMOVE: 'DELETE_PROJECT',
    UPDATE_PROJECT: 'UPDATE_PROJECT',
    SET_PROJECTS: 'SET_PROJECTS',
    SELECT_PROJECT: 'SELECT_PROJECT',
    CLEAR_PROJECTS: 'CLEAR_PROJECTS',
};

const defaultSelectedProject = new Project('', '', '', '', '', true);

class ProjectsState {
    public projects: Project[] = [];
    public selectedProject: Project = defaultSelectedProject;
}

const {
    FETCH,
    CREATE,
    SELECT,
    UPDATE,
    DELETE,
    CLEAR,
} = PROJECTS_ACTIONS;

const {
    ADD,
    REMOVE,
    UPDATE_PROJECT,
    SET_PROJECTS,
    SELECT_PROJECT,
    CLEAR_PROJECTS,
} = PROJECTS_MUTATIONS;

export function makeProjectsModule(api: ProjectsApi): StoreModule<ProjectsState> {
    return {
        state: new ProjectsState(),
        mutations: {
            [ADD](state: any, createdProject: Project): void {
                state.projects.push(createdProject);
            },
            [SET_PROJECTS](state: any, projects: Project[]): void {
                state.projects = projects;

                if (!state.selectedProject.id) {
                    return;
                }

                const projectsCount = state.projects.length;

                for (let i = 0; i < projectsCount; i++) {
                    const project = state.projects[i];

                    if (project.id !== state.selectedProject.id) {
                        continue;
                    }

                    state.selectedProject = project;

                    return;
                }

                state.selectedProject = defaultSelectedProject;
            },
            [SELECT_PROJECT](state: any, projectID: string): void {
                const selected = state.projects.find((project: any) => project.id === projectID);

                if (!selected) {
                    return;
                }

                state.selectedProject = selected;
            },
            [UPDATE_PROJECT](state: any, updateProjectModel: UpdateProjectModel): void {
                const selected = state.projects.find((project: any) => project.id === updateProjectModel.id);
                if (!selected) {
                    return;
                }

                selected.description = updateProjectModel.description;
            },
            [REMOVE](state: any, projectID: string): void {
                state.projects = state.projects.filter(project => project.id !== projectID);

                if (state.selectedProject.id === projectID) {
                    state.selectedProject = new Project();
                }
            },
            [CLEAR_PROJECTS](state: ProjectsState): void {
                state.projects = [];
                state.selectedProject = defaultSelectedProject;
            },
        },
        actions: {
            [FETCH]: async function ({commit}: any): Promise<Project[]> {
                const projects = await api.get();

                commit(SET_PROJECTS, projects);

                return projects;
            },
            [CREATE]: async function ({commit}: any, createProjectModel: CreateProjectModel): Promise<Project> {
                const project = await api.create(createProjectModel);

                commit(ADD, project);

                return project;
            },
            [SELECT]: function ({commit}: any, projectID: string): void {
                commit(SELECT_PROJECT, projectID);
            },
            [UPDATE]: async function ({commit}: any, updateProjectModel: UpdateProjectModel): Promise<void> {
                await api.update(updateProjectModel.id, updateProjectModel.description);

                commit(UPDATE_PROJECT, updateProjectModel);
            },
            [DELETE]: async function ({commit}: any, projectID: string): Promise<void> {
                await api.delete(projectID);

                commit(REMOVE, projectID);
            },
            [CLEAR]: function({commit}: any): void {
                commit(CLEAR_PROJECTS);
            },
        },
        getters: {
            projects: (state: any): Project[] => {
                return state.projects.map((project: any) => {
                    if (project.id === state.selectedProject.id) {
                        project.isSelected = true;
                    }

                    return project;
                });
            },
            selectedProject: (state: any): Project => state.selectedProject,
        },
    };
}
