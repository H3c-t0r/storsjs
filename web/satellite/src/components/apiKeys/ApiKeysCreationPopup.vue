// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="new-api-key" v-if="isPopupShown">
        <h2 class="new-api-key__title">Name Your API Key</h2>
        <HeaderlessInput
            @setData="onChangeName"
            :error="errorMessage"
            placeholder="Enter API Key Name"
            class="full-input"
            width="100%"
        />
        <VButton
            class="next-button"
            label="Next >"
            width="128px"
            height="48px"
            :on-press="onNextClick"
        />
        <div class="new-api-key__close-cross-container" @click="onCloseClick">
            <svg width="16" height="16" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
                <path class="close-cross-svg-path" d="M15.7071 1.70711C16.0976 1.31658 16.0976 0.683417 15.7071 0.292893C15.3166 -0.0976311 14.6834 -0.0976311 14.2929 0.292893L15.7071 1.70711ZM0.292893 14.2929C-0.0976311 14.6834 -0.0976311 15.3166 0.292893 15.7071C0.683417 16.0976 1.31658 16.0976 1.70711 15.7071L0.292893 14.2929ZM1.70711 0.292893C1.31658 -0.0976311 0.683417 -0.0976311 0.292893 0.292893C-0.0976311 0.683417 -0.0976311 1.31658 0.292893 1.70711L1.70711 0.292893ZM14.2929 15.7071C14.6834 16.0976 15.3166 16.0976 15.7071 15.7071C16.0976 15.3166 16.0976 14.6834 15.7071 14.2929L14.2929 15.7071ZM14.2929 0.292893L0.292893 14.2929L1.70711 15.7071L15.7071 1.70711L14.2929 0.292893ZM0.292893 1.70711L14.2929 15.7071L15.7071 14.2929L1.70711 0.292893L0.292893 1.70711Z" fill="#384B65"/>
            </svg>
        </div>
        <div class="blur-content"></div>
    </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';

import HeaderlessInput from '@/components/common/HeaderlessInput.vue';
import VButton from '@/components/common/VButton.vue';

import { ApiKey } from '@/types/apiKeys';
import { API_KEYS_ACTIONS, NOTIFICATION_ACTIONS } from '@/utils/constants/actionNames';

const CREATE = API_KEYS_ACTIONS.CREATE;

@Component({
    components: {
        HeaderlessInput,
        VButton,
    },
})
export default class ApiKeysCreationPopup extends Vue {
    @Prop({default: false})
    private readonly isPopupShown: boolean;

    private name: string = '';
    private errorMessage: string = '';
    private isLoading: boolean = false;
    private key: string = '';

    private FIRST_PAGE = 1;

    public onChangeName(value: string): void {
        this.name = value.trim();
        this.errorMessage = '';
    }

    public onCloseClick(): void {
        this.onChangeName('');
        this.$emit('closePopup');
    }

    public async onNextClick(): Promise<void> {
        if (this.isLoading) {
            return;
        }

        if (!this.name) {
            this.errorMessage = 'API Key name can`t be empty';

            return;
        }

        this.isLoading = true;

        let createdApiKey: ApiKey;

        try {
            createdApiKey = await this.$store.dispatch(CREATE, this.name);
        } catch (error) {
            this.$store.dispatch(NOTIFICATION_ACTIONS.ERROR, error.message);
            this.isLoading = false;

            return;
        }

        this.$store.dispatch(NOTIFICATION_ACTIONS.SUCCESS, 'Successfully created new api key');
        this.key = createdApiKey.secret;
        this.isLoading = false;

        try {
            this.$store.dispatch(API_KEYS_ACTIONS.FETCH, this.FIRST_PAGE);
        } catch (error) {
            await this.$store.dispatch(NOTIFICATION_ACTIONS.ERROR, `Unable to fetch API keys. ${error.message}`);
        }

        this.$emit('closePopup');
        this.$emit('showCopyPopup', this.key);
    }
}
</script>

<style scoped lang="scss">
    .new-api-key {
        padding: 32px 58px 41px 40px;
        background-color: #FFFFFF;
        border-radius: 24px;
        margin-top: 29px;
        max-width: 93.5%;
        height: auto;
        position: relative;

        &__title {
            font-family: 'font_bold';
            font-size: 24px;
            line-height: 29px;
            margin-bottom: 26px;
        }

        .next-button {
            margin-top: 20px;
        }

        &__close-cross-container {
            display: flex;
            justify-content: center;
            align-items: center;
            position: absolute;
            right: 29px;
            top: 29px;
            height: 24px;
            width: 24px;
            cursor: pointer;

            &:hover .close-cross-svg-path {
                fill: #2683FF;
            }
        }

        .blur-content {
            position: absolute;
            top: 100%;
            left: 0;
            background-color: #F5F6FA;
            width: 100%;
            height: 70vh;
            z-index: 100;
            opacity: 0.3;
        }
    }
</style>
