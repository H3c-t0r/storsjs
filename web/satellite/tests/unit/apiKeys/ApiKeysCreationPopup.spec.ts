// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import ApiKeysCreationPopup from '@/components/apiKeys/ApiKeysCreationPopup.vue';
import { ApiKey } from '@/types/apiKeys';
import { makeApiKeysModule } from '@/store/modules/apiKeys';
import { makeProjectsModule } from '@/store/modules/projects';
import { API_KEYS_ACTIONS } from '@/utils/constants/actionNames';
import { Project } from '@/types/projects';
import { ApiKeysApiGql } from '@/api/apiKeys';
import { ProjectsApiGql } from '@/api/projects';

const localVue = createLocalVue();
localVue.use(Vuex);
const apiKeysApi = new ApiKeysApiGql();
const apiKeysModule = makeApiKeysModule(apiKeysApi);
const projectsApi = new ProjectsApiGql();
const projectsModule = makeProjectsModule(projectsApi);

const selectedProject = new Project();
selectedProject.id = '1';

projectsModule.state.selectedProject = selectedProject;

const CREATE = API_KEYS_ACTIONS.CREATE;
const store = new Vuex.Store({modules: { projectsModule, apiKeysModule }});

describe('ApiKeysCreationPopup', () => {
    let value = 'testValue';

    it('renders correctly', () => {
        const wrapper = mount(ApiKeysCreationPopup, {
            store,
            localVue
        });

        expect(wrapper).toMatchSnapshot();
    });

    it('function onCloseClick works correctly', () => {
        const wrapper = mount(ApiKeysCreationPopup, {
            store,
            localVue,
        });

        wrapper.vm.onCloseClick();

        expect(wrapper.emitted()).toEqual({'closePopup': [[]]});
    });

    it('function onChangeName works correctly', () => {
        const wrapper = mount(ApiKeysCreationPopup, {
            store,
            localVue,
        });

        wrapper.vm.onChangeName(value);

        wrapper.vm.$data.name = value.trim();
        expect(wrapper.vm.$data.name).toMatch('testValue');
        expect(wrapper.vm.$data.errorMessage).toMatch('');
    });

    it('action on onNextClick with no name works correctly', async () => {
        const wrapper = mount(ApiKeysCreationPopup, {
            store,
            localVue,
        });

        wrapper.vm.$data.isLoading = false;
        wrapper.vm.$data.name = '';

        await wrapper.vm.onNextClick();

        expect(wrapper.vm.$data.errorMessage).toMatch('API Key name can`t be empty');
    });

    it('action on onNextClick with name works correctly', async () => {
        let testApiKey = new ApiKey('testId', 'testName', 'testCreatedAt', 'test');

        jest.spyOn(apiKeysApi, 'create').mockReturnValue(
            Promise.resolve(testApiKey));

        const wrapper = mount(ApiKeysCreationPopup, {
            store,
            localVue,
        });

        wrapper.vm.$data.isLoading = false;
        wrapper.vm.$data.name = 'testName';

        wrapper.vm.onNextClick();

        let result = await store.dispatch(CREATE, 'testName');

        expect(wrapper.vm.$data.key).toBe(result.secret);
        expect(wrapper.vm.$data.isLoading).toBe(false);
        expect(wrapper.emitted()).toEqual({'closePopup': [[]], 'showCopyPopup': [['test']]});
    });
});
