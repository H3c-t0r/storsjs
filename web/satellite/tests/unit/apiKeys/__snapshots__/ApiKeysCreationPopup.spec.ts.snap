// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiKeysCreationPopup renders correctly 1`] = ``;
